### PR TITLE
Port WinForms TransferForm to WPF

### LINKED ISSUE #15885: Modernize Transfers-window
This is an ongoing effort to improve the resource usage of Cyberduck on Windows.

Addresses:
- **ToDo**: Collect _all_ handle related tickets here in list.

In switching the UI stack from WinForms (currently) to WPF, we gain the ability to reduce our resource usage dramatically, by not requiring a window handle for each and every single control in the transfers window.
Instead there is only one window handle allocated for the entire window, everything else is done in software.
A welcome side-effect of this is the offload of rendering to the GPU, reducing the software-rendering requirements of WinForms.

With this, the foundation (controller-architecture) is switching to a view-model based approach, allowing an easier transition to another UI framework in the future.

WPF Limitations as outlined in [#13036](https://redirect.github.com/iterate-ch/cyberduck/issues/13036) apply.

**Workaround**
> The [Microsoft documentation](https://learn.microsoft.com/en-us/windows/win32/sysinfo/user-objects) provides more information on this issue:
>
> User interface objects support only one handle per object. Processes cannot inherit or duplicate handles to user objects. Processes in one session cannot reference a user handle in another session.
>
> There is a theoretical limit of 65,536 user handles per session. However, the maximum number of user handles that can be opened per session is usually lower, since it is affected by available memory. There is also a default per-process limit of user handles.
>
> You can try to solve the issue by increasing the maximum number of user object handles if deleting the Transfers folder doesn't have any effect. Please follow the steps below to do that:
> 
> 1. Use the shortcut Win R to open the search box, type regedit, and open the Registry Editor
> 2. Copy the following line into the address bar
`Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Windows`
> 3. Double-click on the entry USERProcessHandleQuota and increase the maximum number of user handlers. Note: The value must be set to a number between 200 and 18000.
> 4. Restart the device

### PR DESCRIPTION
**Pull request by AliveDevil**
Resolves #15885

### CODE REVIEW

File: core/native/refresh/src/main/csharp/Services/IconCache.cs [hunk 1]
```diff
@@ -106,5 +121,24 @@ public bool TryGetIcon<T>(object Key, int size, out T image, string classifier =
         public ReaderWriterLockSlimExtensions.UpgradeableReadLock UpgradeableReadLock() => writer.UseUpgradeableReadLock();
 
         public ReaderWriterLockSlimExtensions.WriteLock WriteLock() => writer.UseWriteLock();
+
+        private static void OnCacheEvict(CacheEntryRemovedArguments arguments)
```
Comment (AliveDevil) [inline core/native/refresh/src/main/csharp/Services/IconCache.cs:125]: Used to temporarily store extracted file icons (which will be garbage collected once not used in cache anymore).

File: core/native/refresh/src/main/csharp/Services/IconProvider.Path.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+﻿using ch.cyberduck.core;
+using System;
+
+namespace Ch.Cyberduck.Core.Refresh.Services
+{
+    public partial class IconProvider<T>
+    {
+        public T GetPath(Path path, int size)
```
Comment (AliveDevil) [inline core/native/refresh/src/main/csharp/Services/IconProvider.Path.cs:8]: Moved from WinFormsIconProvider to base - previously used in ProgressController.

File: core/native/refresh/src/main/csharp/Services/IconProvider.cs [hunk 3]
```diff
@@ -105,17 +105,21 @@ public T GetApplication(Application application, int size)
         public T GetFileIcon(string filename, bool isFolder, bool large, bool isExecutable)
         {
             string key = string.Empty;
+            string fileInfo = filename;
```
Comment (AliveDevil) [inline core/native/refresh/src/main/csharp/Services/IconProvider.cs:108]: Changed logic here to be able to extract local file icons correctly, and not pollute the cache with wrong information.
e.g. calling GetFileIcon("C:\A\B.exe") would pollute cache key "ext:exe" to forever use that icon.

File: core/native/refresh/src/main/csharp/Services/WpfIconProvider.cs [hunk 4]
```diff
@@ -137,6 +73,36 @@ protected override BitmapSource NearestFit(IEnumerable<BitmapSource> sources, in
             return nearestFit;
         }
 
+        protected override BitmapSource Overlay(BitmapSource baseImage, BitmapSource overlay, int size)
+        {
+            DrawingVisual visual = new();
+            using (var context = visual.RenderOpen())
```
Comment (AliveDevil) [inline core/native/refresh/src/main/csharp/Services/WpfIconProvider.cs:79]: Unfortunately required.

File: core/src/main/csharp/ch/cyberduck/core/CollectionObservables.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+#nullable enable
+
+namespace ch.cyberduck.core;
+
+public static class CollectionObservables
+{
+    public delegate TResult ObservableTransform<TSource, TResult>(TSource obj, TResult? cached);
+
+    public static IObservable<IChangeSet<TResult>> ToObservableChangeSet<TSource, TResult>(
```
Comment (AliveDevil) [inline core/src/main/csharp/ch/cyberduck/core/CollectionObservables.cs:14]: Create Observable from ch.cyberduck.core.Collection
Basically allows watching for collectionX-notifications

File: windows/src/main/csharp/NativeMethods.txt [hunk 6]
```diff
@@ -27,6 +28,7 @@ RegisterClipboardFormat
 SB_BOTTOM
 SetMenuInfo
 SetMenuItemInfo
+SetThreadDpiAwarenessContext
```
Comment (AliveDevil) [inline windows/src/main/csharp/NativeMethods.txt:31]: Required to get the secondary STA UI thread to follow DPI

File: windows/src/main/csharp/XamlGeneratedNamespace/GeneratedApplication.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Threading;
+using static Windows.Win32.PInvoke;
+
+namespace XamlGeneratedNamespace;
+
+partial class GeneratedApplication
+{
+    public static GeneratedApplication OffThread()
```
Comment (AliveDevil) [inline windows/src/main/csharp/XamlGeneratedNamespace/GeneratedApplication.cs:23]: Move Transfer-window (or rather: All (future) Windows, that aren't WinForms) into separate thread

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/SynchronizedObservableObject.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+using CommunityToolkit.Mvvm.ComponentModel;
+using System.ComponentModel;
+using System.Threading;
+
+namespace ch.cyberduck.ui.ViewModels;
+
+public abstract class SynchronizedObservableObject : ObservableObject
+{
+    private readonly SynchronizationContext sync = SynchronizationContext.Current;
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/SynchronizedObservableObject.cs:22]: Capturing creating synchronization context to schedule PropertyChanged notifications back to UI thread

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        {
+            { } state => state.Message,
+            _ => Transfer.Timestamp switch
+            {
+                { } date => dateFormatter.getLongFormat(date.getTime(), false),
+                _ => null,
+            },
+        };
+
+        public string Name => Transfer.Name;
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs:79]: Accessibility Help Text, invoked on Windows using `{Narrator-Key}+0`

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+        public TransferViewModel(TransferController controller, TransferModel transfer, Locale locale)
+        {
+            this.locale = locale;
+            Transfer = transfer;
+
+            transferTypeName = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(Transfer.Model.getType().name());
+
+            var progressState = Observable.Create<TransferProgressModel>(observer =>
+            {
+                bool emitEmpty = true;
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs:129]: Cannot work around. The Controller.Progress is an ObservableCache, which allows removal/adding of new entries.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+
+        partial void OnProgressStateChanged(TransferProgressModel value)
+        {
+            if (value is null)
+            {
+                Running = false;
+            }
+            else
+            {
+                Running = true;
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs:192]: Update several UI states

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+                OnPropertyChanged(nameof(ItemStatus));
+                OnPropertyChanged(nameof(MessageText));
+                OnPropertyChanged(nameof(ProgressText));
+            }
+        }
+
+        partial void OnProgressStateChanged(TransferProgressModel old, TransferProgressModel value)
+        {
+            progressStateNotifications.Disposable = value?.WhenAnyPropertyChanged().Subscribe(OnProgressStateChanged);
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs:207]: SerialDisposable disposes previous disposable, then assigns new value.
This way we don't need to track the event-subscription manually.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        private void OnTransferChanged(object sender, PropertyChangedEventArgs e)
+        {
+            switch (e.PropertyName)
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransferViewModel.cs:212]: Make UI aware of changes to computed properties.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+            if (SetProperty(ref selectedBandwidth, value))
+            {
+                OnSelectedBandwidthChanged(value);
+            }
+        }
+    }
+
+    public ReadOnlyObservableCollection<TransferViewModel> SelectedTransfers => selectedTransfers;
+
+    public bool ToolbarCleanup
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:84]: Maybe move away from here, as it doesn't belong here, but as the preferences aren't reactive/observable/INPC-capable, properties are stored here for the time being.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+        set
+        {
+            OnPropertyChanging();
+            preferences.setProperty("transfer.toolbar.trash", value);
+            OnPropertyChanged();
+            OnPropertyChanged(nameof(ToolbarSeparatorLeft));
+        }
+    }
+
+    public bool ToolbarSeparatorLeft
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:180]: Ugly, but extracting a TransferToolbarButtonProvider which dynamically changes the toolbar in the Transfers window would be more complicated.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,535 @@
...
+        Preferences preferences,
+        WpfIconProvider iconProvider)
+    {
+        this.controller = controller;
+        this.store = store;
+        this.preferences = preferences;
+        this.iconProvider = iconProvider;
+
+        var localTransfers = store.Transfers.Connect().ObserveOnDispatcher()
+            .Transform(m => new TransferViewModel(controller, m, locale)).DisposeMany()
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:215]: Store has dynamic list of TransferModel, which is transformed here to TransferViewModel (on Dispatcher, thus captured SynchronizationContext of SynchronizingObservableObject is set to UI)

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 17]
```diff
@@ -0,0 +1,535 @@
...
+            .Bind(out transfers)
+            .AutoRefresh(m => m.IsSelected).Filter(m => m.IsSelected).Bind(out selectedTransfers)
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:217]: All TransferViewModels have IsSelected-property, automatically filter and store in a cache for future usage.
WPF doesn't allow binding to Selector.SelectedItems (this is a workaround)

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 18]
```diff
@@ -0,0 +1,535 @@
...
+            .AsObservableCache();
+        localTransfers.Connect().WhenAnyPropertyChanged().Subscribe(OnSelectedTransferPropertyChanged);
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:219]: Fastest way to get any item in that collection to force a UI update

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+        if (!connectionLookup.TryGetValue(rate, out var selection))
+        {
+            connectionLookup[rate] = selection = new(rate, false, null);
+            Connections.Add(selection);
+        }
+
+        return selection;
+    }
+
+    [RelayCommand]
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:261]: Mentioned once: RelayCommandAttribute from CommunityToolkit.Mvvm is a source generator, wiring up a lazy RelayCommand, which invokes OnClean in Execute, and may call CanExecute-parameter for validation whether the command should run.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,535 @@
...
+                TransferOptions options = new TransferOptions().resume(false).reload(true);
+                controller.StartTransfer(item.Transfer.Model, options);
+            }
+        }
+    }
+
+    [RelayCommand(CanExecute = nameof(ValidateRemoveCommand))]
+    private void OnRemove()
+    {
+        using var copy = SelectedTransfers.CopyToPool(true);
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:293]: While we are removing transfers, these are _immediately_ synced back to SelectedTransfers (due to them being removed), which would invalidate the enumerator, thus create temporary pooled copy.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 21]
```diff
@@ -0,0 +1,535 @@
...
+                action.getMeter().reset();
+            }
+
+            transfer.Transfer.Refresh();
+        }
+    }
+
+    private void OnSelectedTransferPropertyChanged(TransferViewModel _)
+    {
+        OpenCommand.NotifyCanExecuteChanged();
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:336]: Need to explicitely state when commands CanExecute-method may have changed.

File: windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs [hunk 22]
```diff
@@ -0,0 +1,535 @@
...
+        if (SelectedBandwidth?.Enabled == false)
+        {
+            bandwidthLookup.Remove(SelectedBandwidth.Bandwidth);
+            Bandwidth.Remove(SelectedBandwidth);
+        }
+
+        var rate = (int)SelectedTransfer.Transfer.BandwidthRate;
+        if (!bandwidthLookup.TryGetValue(rate, out var selection))
+        {
+            bandwidthLookup[rate] = selection = new(rate, false, BandwidthProvider.Format(rate));
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/ViewModels/TransfersViewModel.cs:411]: Adding a fallback element with the currently configured rate.
This element allows for changed values using default.properties, or editing user.config.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersItem.xaml [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+                      Margin="0,3"
+                      VerticalAlignment="Center"
+                      Grid.Column="1"
+                      Grid.Row="0"
+                      VirtualizingPanel.IsVirtualizing="True"
+                      VirtualizingPanel.VirtualizationMode="Recycling" />
+
+            <Grid Grid.Column="1"
+                  Grid.Row="1">
+                <ProgressBar Value="{Binding Progress}"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersItem.xaml:49]: Two controls stacked ontop each other.
ProgressBar shown when Running property is true
TextBlock shown when Running property is false.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml [hunk 24]
```diff
@@ -0,0 +1,373 @@
...
+    <DockPanel>
+        <DockPanel Name="ToolBar"
+                   Background="Transparent"
+                   Margin="8,3"
+                   LastChildFill="False"
+                   DockPanel.Dock="Top"
+                   KeyboardNavigation.TabIndex="0"
+                   KeyboardNavigation.TabNavigation="Local">
+            <DockPanel.Resources>
+                <Style TargetType="Button"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml:36]: Copying Toolbar styles for Buttons and Separators.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml [hunk 25]
```diff
@@ -0,0 +1,373 @@
...
+        </Style>
+    </Window.Resources>
+
+    <DockPanel>
+        <DockPanel Name="ToolBar"
+                   Background="Transparent"
+                   Margin="8,3"
+                   LastChildFill="False"
+                   DockPanel.Dock="Top"
+                   KeyboardNavigation.TabIndex="0"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml:33]: We want keyboard navigation to be:
Toolbar, ListBox, Border (for URL and Local labels).

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml [hunk 26]
```diff
@@ -0,0 +1,373 @@
...
+                              StaysOpenOnClick="True"
+                              IsCheckable="True" />
+                    <MenuItem Header="{xaml:Text Show,Table=Localizable}"
+                              IsChecked="{Binding ToolbarShow}"
+                              StaysOpenOnClick="True"
+                              IsCheckable="True" />
+                </ContextMenu>
+            </DockPanel.ContextMenu>
+
+            <StackPanel Orientation="Horizontal"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml:81]: This is ugly, but WPF doesn't have an AppBar, or any way of making the toolbar not suck.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml [hunk 27]
```diff
@@ -0,0 +1,373 @@
...
+                               Height="24"
+                               Width="24"
+                               Style="{StaticResource ButtonImage}" />
+                        <TextBlock Name="ReloadButtonLabel"
+                                   Text="{xaml:Text Reload,Table=Localizable}" />
+                    </StackPanel>
+                </Button>
+            </StackPanel>
+
+            <StackPanel Orientation="Horizontal"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml:168]: WPF toolbar doesn't allow start/end alignment of elements. So use two StackPanels docked left/right.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+                           Margin="3"
+                           TextTrimming="CharacterEllipsis"
+                           VerticalAlignment="Center"
+                           Grid.Column="2"
+                           Grid.Row="1"
+                           AutomationProperties.HelpText="{Binding Text,ElementName=LocalFileLabel}" />
+            </Grid>
+        </Border>
+
+        <ListBox ItemsSource="{Binding Transfers}"
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml:339]: Last item in DockPanel is Filling by default.

File: windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml.cs [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+    public partial class TransfersWindow : Window
+    {
+        public TransfersWindow()
+        {
+            InitializeComponent();
+        }
+
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            ToolBar.Measure(new(double.PositiveInfinity, double.PositiveInfinity));
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/Views/TransfersWindow.xaml.cs:28]: Force window to not be smaller than the minimum required size by dockpanel.
The make-shift Toolbar doesn't support item collapsing/overflow.

File: windows/src/main/csharp/ch/cyberduck/ui/controller/BrowserController.cs [hunk 30]
```diff
@@ -2281,7 +2283,10 @@ private void View_BrowserCanDrop(OlvDropEventArgs args)
 
         private void View_ShowTransfers()
         {
-            ITransferView view = TransferController.Instance.View;
+            TransferController.Instance.ShowWindow();
+            //var view = new TransfersWindow();
+            //view.Show();
+            /*ITransferView view = TransferController.Instance.View;
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/controller/BrowserController.cs:2289]: ToDo: Bring back move to current desktop

File: windows/src/main/csharp/ch/cyberduck/ui/controller/MainController.cs [hunk 31]
```diff
@@ -848,7 +856,7 @@ private void InitializeTransfers()
                 _bc.Invoke(() =>
                 {
                     transfersSemaphore.Wait();
-                    TransferController.Instance.View.Show();
+                    TransferController.Instance.ShowWindow();
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/controller/MainController.cs:859]: View is lazily created if required. Use instance method which handles everything (including dispatching to WPF secondary thread).

File: windows/src/main/csharp/ch/cyberduck/ui/controller/PromptLoginController.cs [hunk 32]
```diff
@@ -56,22 +58,26 @@ public PromptLoginController(WindowController c) : base(c)
...
-                            // Never show again.
-                            PreferencesFactory.get().setProperty(preference, true);
-                        }
-                        switch (option)
-                        {
-                            case 1:
-                                throw new LoginCanceledException();
-                        }
-                    });
+                var result = UiUtils.Show(
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/controller/PromptLoginController.cs:61]: IWindowController doesn't have CommandBox.
Use readily available Utils-helper methods.

File: windows/src/main/csharp/ch/cyberduck/ui/controller/TransferController.cs [hunk 33]
```diff
@@ -16,731 +16,330 @@
...
+                return;
             }
-        }
 
-        private void View_TranscriptHeightChangedEvent()
-        {
-            _preferences.setProperty("queue.transcript.size.height", View.TranscriptHeight);
+            if (wait)
+            {
+                MainController.Application.SynchronizationContext.Send(
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/controller/TransferController.cs:114]: In order to not get bitten by not-supported UI display scaling when opening prompts from the Transfer-window thread which has DPI display scaling enabled, schedule everything from background actions to WinForms main thread.

File: windows/src/main/csharp/ch/cyberduck/ui/controller/TransferController.cs [hunk 34]
```diff
@@ -16,731 +16,330 @@
...
 
-                }
-                else
-                {
-                    list.Add(new KeyValuePair<int, string>(connections,
-                        MessageFormat.format(LocaleFactory.localizedString("{0} Connections", "Transfer"), connections)));
-                }
+            if (!transfers.contains(transfer))
+            {
+                /* TODO: Queue Size Warn */
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/controller/TransferController.cs:167]: TODO!

File: windows/src/main/csharp/ch/cyberduck/ui/core/TaskbarApplicationBadgeLabeler.cs [hunk 35]
```diff
@@ -28,7 +28,7 @@ public sealed class TaskbarApplicationBadgeLabeler : ApplicationBadgeLabeler
     {
         public void badge(string text)
         {
-            if (Cyberduck.Core.Utils.IsBlank(text))
+            /*if (Cyberduck.Core.Utils.IsBlank(text))
```
Comment (AliveDevil) [inline windows/src/main/csharp/ch/cyberduck/ui/core/TaskbarApplicationBadgeLabeler.cs:31]: TODO

**Review by ylangisc — APPROVED** [review]

**Pull request merged**